shared/src/publication/Manifest.ts

```diff
@@ -0,0 +1,232 @@
+    json.readingOrder = this.readingOrder.serialize();
+    if (this.resources) json.resources = this.resources.serialize();
+    if (this.tableOfContents) json.toc = this.tableOfContents.serialize();
+    PublicationCollection.serializeCollection(json, this.subcollections);
+    return json;
+  }
+
+  /** Finds the first link with the given relation in the manifest's links. */
+  public linkWithRel(rel: string): Link | undefined {
+    const links = new Array<Links>();
+    links.push(this.readingOrder);
+    if (this.resources) {
+      links.push(this.resources);
+    }
+    links.push(this.links);
+
+    let result: Link | undefined;
+
+    for (const collection of links) {
+      result = collection.findWithRel(rel);
+      if (result !== undefined) {
+        return result;
+      }
+    }
+
+    return result;
+  }
+
+  /** Finds all the links with the given relation in the manifest's links. */
+  public linksWithRel(rel: string): Array<Link> {
+    const result = [];
+    result.push(this.readingOrder.filterByRel(rel));
+    if (this.resources) {
+      result.push(this.resources.filterByRel(rel));
+    }
+    result.push(this.links.filterByRel(rel));
+
+    return result.reduce((acc, val) => acc.concat(val), []);
+  }
+
+  /** Finds the first Link having the given `href` in the publication's links. */
+  public linkWithHref(href: string): Link | undefined {
+    const find = (links: Array<Links>): Link | undefined => {
+      let result = undefined;
+
+      for (const collection of links) {
+        result = collection.findWithHref(href);
+        if (result !== undefined) {
+          return result;
+        }
+      }
+
+      const children = new Array<Links>();
+
+      links.forEach(item => {
+        const arr = [];
+        for (const link of item.items) {
+          if (link.alternates) {
+            arr.push(link.alternates);
+          }
+          if (link.children) {
+            arr.push(link.children);
+          }
+        }
+        children.push(...arr);
+      });
+
+      if (children.length > 0) {
+        result = find(children);
+      }
+
+      return result;
+    };
+
+    const links: Array<Links> = [];
+    links.push(this.readingOrder);
+    if (this.resources) {
+      links.push(this.resources);
+    }
+    links.push(this.links);
+
+    const link = find(links);
+
+    if (link !== undefined) {
+      return link;
+    }
+
+    const parts = href.split(/[#]/);
+
+    if (parts.length < 2) return;
+
+    const shortHref = parts[0];
+
+    return this.linkWithHref(shortHref);
+  }
+
+  /** The URL where this publication is served, computed from the `Link` with `self` relation.
+   *  e.g. https://provider.com/pub1293/manifest.json gives https://provider.com/pub1293/
+   */
+  public get baseURL(): string | undefined {
+    const selfLink = this.links.items.find(
+      el => el.rels && el.rels.has('self')
+    );
+    if (selfLink) {
+      let href = selfLink.href;
```

Comment:
Remove the `if (selfLink) {` statement and just have:

```suggestion
      let href = selfLink?.href;
```